packages/eslint-plugin/src/rules/prefer-find.ts

```diff
@@ -0,0 +1,110 @@
+import type { TSESLint } from '@typescript-eslint/utils';
+import { AST_NODE_TYPES } from '@typescript-eslint/utils';
+import * as tsutils from 'ts-api-utils';
+
+import { createRule, getParserServices } from '../util';
+
+export default createRule({
+  name: 'prefer-find',
+  meta: {
+    docs: {
+      description:
+        'Enforce the use of Array.prototype.find() over Array.prototype.filter() followed by [0] when looking for a single result',
+      requiresTypeChecking: true,
+      recommended: 'stylistic',
+    },
+    fixable: 'code',
+    messages: {
+      preferFind: 'Use .filter(...)[0] instead of .find(...)',
+      preferFindFix: 'Use .filter(...)[0] instead of .find(...)',
+    },
+    schema: [],
+    type: 'suggestion',
+    hasSuggestions: true,
+  },
+
+  defaultOptions: [],
+
+  create(context) {
+    const services = getParserServices(context);
+    const checker = services.program.getTypeChecker();
+
+    return {
+      MemberExpression(node): void {
+        // Check if it looks like <<stuff>>[0] or <<stuff>>['0'], but not <<stuff>>?.[0]
+        if (
+          node.computed &&
+          !node.optional &&
+          node.property.type === AST_NODE_TYPES.Literal &&
+          (node.property.value === 0 ||
+            node.property.value === '0' ||
+            node.property.value === 0n)
+        ) {
+          // Check if it looks like <<stuff>>(...)[0], but not <<stuff>>?.(...)[0]
+          if (
+            node.object.type === AST_NODE_TYPES.CallExpression &&
+            !node.object.optional
```

Comment:
[Question] Why not? This seems like it'd be a case folks would want to hit. We'd probably get a bug report from it not working.

E.g. for the `([]?.filter(f))[0];` case, I don't see why we wouldn't want to flag that.